File: extensions/interactions/NumberWithUnitInput/NumberWithUnitInput.py
# coding: utf-8
#
# Copyright 2014 The Oppia Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, softwar
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from extensions.interactions import base


class NumberWithUnitInput(base.BaseInteraction):
    """Interaction for number with unit input."""

    name = 'Number with Unit'
    category = 'Science'
    description = (
    'Allows learners to enter integers and floating point numbers followed by standard units.')
    display_mode = base.DISPLAY_MODE_INLINE
    dependency_ids = []
    _handlers = [{
    'name': 'submit', 'obj_type': 'RealWithUnit'}]

    # _customization_arg_specs = [] # Not sure about this one yet.

File: extensions/interactions/NumberWithUnitInput/NumberWithUnitInput.js
// Copyright 2014 The Oppia Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS-IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


/**
 * Directive for the NumberWithUnitInput interaction.
 *
 * IMPORTANT NOTE: The naming convention for customization args that are passed
 * into the directive is: the name of the parameter, followed by 'With',
 * followed by the name of the arg.
 */

oppia.directive('oppiaInteractiveNumberWithUnitInput', [function() {
  return {
    restrict: 'E',
    scope: {},
    templateUrl: 'interaction/NumberWithUnitInput',
    controller: ['$scope', '$attrs', function($scope, $attrs) {
      $scope.labelForFocusTarget = $attrs.labelForFocusTarget || null;
      $scope.errorMessage = '';

      $scope.NUMBER_WITH_UNIT_INPUT_FORM_SCHEMA = {
        type: 'unicode',
        ui_config: {}
      };

      $scope.preprocess = function(unprocessedAnswer) {
        var preprocessedAnswer = unprocessedAnswer.replace(/\s/g, '').replace(
          ',', '');
        return preprocessedAnswer;
      }

      $scope.postprocess = function(answer, parsedAnswer) {
        var processedAnswer = {'raw': answer, 'parsed': parsedAnswer}
        return processedAnswer;
      }

      $scope.submitAnswer = function(answer) {
        if (answer) {
          var strippedAnswer = $scope.preprocess(answer);
          try {
            var parsedAnswer = numberWithUnitInputParser.parse(strippedAnswer);
          } catch(error) {
            if (error.message) {
              $scope.errorMessage = error.message;
              return;
            } else {
              $scope.errorMessage = 'Please enter a number with units.';
              throw error;
            }
          }

          var processedAnswer = $scope.postprocess(answer, parsedAnswer);

          $scope.$parent.$parent.submitAnswer(processedAnswer, 'submit');
        }
      }
    }]
  }
}]);

oppia.directive('oppiaResponseNumberWithUnitInput', [
    'oppiaHtmlEscaper', function(oppiaHtmlEscaper) {
  return {
    scope: {},
    templateUrl: 'response/NumberWithUnitInput',
    controller: ['$scope', '$attrs', function($scope, $attrs) {
      $scope.answer = oppiaHtmlEscaper.escapedJsonToObj($attrs.answer);
    }]
  }
}]);
